R/derive_param_tte.R

```diff
@@ -0,0 +1,616 @@
+  assert_list_element(
+    list = event_conditions,
+    element = "dataset_name",
+    condition = dataset_name %in% source_names,
+    source_names = source_names,
+    message_text = paste0(
+      "The dataset names must be included in the list specified for the ",
+      "`source_datasets` parameter.\n",
+      "Following names were provided by `source_datasets`:\n",
+      enumerate(source_names, quote_fun = squote)
+    )
+  )
+  assert_list_element(
+    list = censor_conditions,
+    element = "dataset_name",
+    condition = dataset_name %in% source_names,
+    source_names = source_names,
+    message_text = paste0(
+      "The dataset names must be included in the list specified for the ",
+      "`source_datasets` parameter.\n",
+      "Following names were provided by `source_datasets`:\n",
+      enumerate(source_names, quote_fun = squote)
+    )
+  )
+  assert_logical_scalar(create_datetime)
+  assert_varval_list(set_values_to, accept_expr = TRUE, optional = TRUE)
+  if (!is.null(set_values_to$PARAMCD) & !is.null(dataset)) {
+    assert_param_does_not_exist(dataset, quo_get_expr(set_values_to$PARAMCD))
+  }
+  if (!is.null(by_vars)) {
+    # ensure that the by variables are contained in all source datasets #
+    by_vars_chr <- vars2chr(by_vars)
+    # determine which source datasets need to be extended #
+    by_groups <- list()
+    extend <- vector("list", length(source_datasets))
+    for (i in seq_along(source_datasets)) {
+      missing_by_vars <- setdiff(by_vars_chr, names(source_datasets[[i]]))
+      if (length(missing_by_vars) == 0) {
+        # all by variables are included in the source dataset #
+        extend[[i]] <- FALSE
+        by_groups <-
+          append(by_groups,
+                 list(unique(select(
+                   source_datasets[[i]], !!!by_vars
+                 ))))
+      }
+      else if (!setequal(by_vars_chr, missing_by_vars)) {
+        # only some of the by variables are included in the source dataset #
+        abort(paste("Only",
+                    paste(setdiff(by_vars_chr, missing_by_vars), collapse = ", "),
+                    "are included in source dataset",
+                    names(source_datasets)[[i]],
+                    ".\n The source dataset must include all or none of the by variables."))
+      }
+      else {
+        extend[[i]] <- TRUE
+      }
+    }
+    if (length(by_groups) == 0) {
+      abort(paste0("The by variables (",
+                   paste(by_vars_chr, collapse = ", "),
+                   ") are not contained in any of the source datasets."))
+    }
+    # extend source datasets #
+    by_groups <- unique(bind_rows(by_groups))
+    for (i in seq_along(source_datasets)) {
+      if (extend[[i]]) {
+        source_datasets[[i]] <-
+          full_join(mutate(by_groups, temp_dummy = 1),
+                    mutate(source_datasets[[i]], temp_dummy = 1),
+                    by = "temp_dummy") %>%
+          select(-temp_dummy)
+      }
+    }
+  }
```

Comment:
Let's put this into a function which we just call here. This kind of distracts from the overall flow and purpose of `derive_param_tte()`.